apollo-router/src/plugins/record_replay/record.rs

```diff
@@ -0,0 +1,342 @@
+        let storage_path = PathBuf::from(
+            init.config
+                .storage_path
+                .unwrap_or_else(default_storage_path),
+        );
+
+        write_file(
+            storage_path.clone().into(),
+            "README.md",
+            include_str!("recording-readme.md").as_bytes(),
+        )
+        .await?;
+
+        let plugin = Self {
+            enabled: init.config.enabled,
+            supergraph_sdl: init.supergraph_sdl.clone(),
+            storage_path: storage_path.into(),
+            schema: Arc::new(Schema::parse(
+                init.supergraph_sdl.clone().as_str(),
+                &Configuration::default(),
+            )?),
+        };
+
+        Ok(plugin)
+    }
+
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        if !self.enabled {
+            return service;
+        }
+
+        let dir = self.storage_path.clone();
+        let supergraph_sdl = self.supergraph_sdl.clone();
+
+        ServiceBuilder::new()
+            .map_future(move |future| {
+                let dir = dir.clone();
+                let supergraph_sdl = supergraph_sdl.clone();
+
+                async move {
+                    let res: router::Response = future.await?;
+                    let (parts, stream) = res.response.into_parts();
+
+                    let headers = parts.headers.clone();
+                    let context = res.context.clone();
+
+                    let after_complete = once(async move {
+                        if recording_enabled(&context) {
+                            let client_request =
+                                context.get::<_, RequestDetails>(CLIENT_REQUEST)?;
+                            let client_response =
+                                context.get::<_, ResponseDetails>(CLIENT_RESPONSE)?;
+
+                            if let (Some(client_request), Some(client_response)) =
+                                (client_request, client_response)
+                            {
+                                let res_headers = externalize_header_map(&headers)?;
+                                let client_response = ResponseDetails {
+                                    headers: res_headers,
+                                    ..client_response
+                                };
+
+                                let operation_name = client_request
+                                    .operation_name
+                                    .clone()
+                                    .unwrap_or("UnnamedOperation".to_string());
+
+                                let unix_time = std::time::SystemTime::now()
+                                    .duration_since(std::time::UNIX_EPOCH)?
+                                    .as_secs();
+
+                                let filename =
+                                    format!("{}-{}.json", operation_name, unix_time).to_string();
```

Comment:
I think in the future it might be better to hash the query and use that hash in the filename instead of unix_time. If we do that then for the same operation/request it will override the previous file

Reply:
i'm not sure i want to overwrite files — i guess i could if i included other aspects of the request (variables, headers) in the hash too

Reply:
done!